src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -90,72 +94,84 @@ internal Type GetTypeFromCoreAssembly(CoreTypeId typeId)
             return null;
         }
 
-        internal void AppendMetadata(MetadataBuilder metadata)
+        internal void AppendMetadata()
         {
             // Add module metadata
-            metadata.AddModule(
+            _metadataBuilder.AddModule(
                 generation: 0,
-                moduleName: metadata.GetOrAddString(_name),
-                mvid: metadata.GetOrAddGuid(Guid.NewGuid()),
+                moduleName: _metadataBuilder.GetOrAddString(_name),
+                mvid: _metadataBuilder.GetOrAddGuid(Guid.NewGuid()),
                 encId: default,
                 encBaseId: default);
 
             // Create type definition for the special <Module> type that holds global functions
-            metadata.AddTypeDefinition(
+            _metadataBuilder.AddTypeDefinition(
                 attributes: default,
                 @namespace: default,
-                name: metadata.GetOrAddString("<Module>"),
+                name: _metadataBuilder.GetOrAddString("<Module>"),
                 baseType: default,
                 fieldList: MetadataTokens.FieldDefinitionHandle(1),
                 methodList: MetadataTokens.MethodDefinitionHandle(1)); ;
 
             // Add each type definition to metadata table.
-            foreach (TypeBuilderImpl typeBuilder in _typeDefStore)
+            foreach (TypeBuilderImpl typeBuilder in _typeDefinitions)
             {
-                TypeReferenceHandle parent = default;
+                EntityHandle parent = default;
                 if (typeBuilder.BaseType is not null)
                 {
-                    // TODO: need to handle the case when the base is from same assembly
-                    parent = GetTypeReference(metadata, typeBuilder.BaseType);
+                    parent = GetTypeHandle(typeBuilder.BaseType);
                 }
 
-                TypeDefinitionHandle typeDefinitionHandle = MetadataHelper.AddTypeDefinition(metadata, typeBuilder, parent, _nextMethodDefRowId, _nextFieldDefRowId);
+                TypeDefinitionHandle typeDefinitionHandle = MetadataHelper.AddTypeDefinition(_metadataBuilder, typeBuilder, parent, _nextMethodDefRowId, _nextFieldDefRowId);
```

Comment:
`MetadataHelper` looks like unnecessary abstraction. Can all the methods like `AddTypeDefinition`, `AddMethodDefinition`, etc. be private instance methods here?

Reply:
Sure, they can